### PR TITLE
Rolling upgrade azure

### PR DESCRIPTION
**Pull request by KnifeyMoloko**
This PR introduces commits related to adding a rolling upgrade job for the Azure backend, i.e. running the rolling upgrade job on Azure, using the Azure machine images. 
An additional commit was added with a small refactor change for the `rollingUpgradePipeline.groovy` script file.

Git task: https://github.com/scylladb/qa-tasks/issues/336
Staging job for testing: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/ru-azure

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: vars/createSctRunner.groovy [hunk 1]
```diff
@@ -21,6 +21,13 @@ def call(Map params, Integer test_duration, String region) {
     } else {
         availability_zone = params.availability_zone
     }
+
+    if ( params.backend.equals("azure") ) {
```
Comment (vponomaryov) [inline vars/createSctRunner.groovy:25]: Please, add this block of code after current line 33 to keep logical code blocks in one place.

File: vars/rollingUpgradePipeline.groovy [hunk 2]
```diff
@@ -94,22 +100,24 @@ def call(Map pipelineParams) {
...
-                                        base_versions_list,
-                                        pipelineParams.linux_distro,
-                                        params.new_scylla_repo
-                                    )
-                                    (testDuration,
-                                     testRunTimeout,
-                                     runnerTimeout,
-                                     collectLogsTimeout,
-                                     resourceCleanupTimeout) = getJobTimeouts(params, builder.region)
+                    catchError(stageResult: "FAILURE") {
```
Comment (vponomaryov) [inline vars/rollingUpgradePipeline.groovy:103]: Addition of the "catchError" wrapper looks as unrelated to the "azure" parts.
So, better to switch this part to a different commit. And First commit which is "fix(vars): add 'azure' to getCloudProviderFromBackned.groovy map" merge in the main one with all the azure code changes.

File: vars/createSctRunner.groovy (hunk 1 above)
Comment (KnifeyMoloko) [reply vars/createSctRunner.groovy:30]: Fixed - moved it below the availability_zone blocks

File: vars/rollingUpgradePipeline.groovy (hunk 2 above)
Comment (KnifeyMoloko) [reply vars/rollingUpgradePipeline.groovy:103]: Fixed - moved the change to a separate commit. Also merged the getCloudProviderFromBackend with other azure-specific changes in the first commit

File: vars/createSctRunner.groovy [hunk 3]
```diff
@@ -21,20 +21,29 @@ def call(Map params, Integer test_duration, String region) {
...
+    }
+
     println(params)
     sh """
     #!/bin/bash
     set -xe
     env
 
-    if [[ "$cloud_provider" == "aws" || "$cloud_provider" == "gce" || "$cloud_provider" == "azure" ]]; then
+    if [ "$cloud_provider" = "aws" ] || [ "$cloud_provider" = "gce" ] || [ "$cloud_provider" = "azure" ]; then
```
Comment (vponomaryov) [inline vars/createSctRunner.groovy:42]: Why do you need these changes? On line 38 we know for sure that we run it in bash and it must work there.
Why?

File: vars/createSctRunner.groovy [hunk 4]
```diff
@@ -21,20 +21,29 @@ def call(Map params, Integer test_duration, String region) {
     } else {
         availability_zone = params.availability_zone
     }
+
```
Comment (vponomaryov) [inline vars/createSctRunner.groovy:24]: Do you really need these 2 new lines here?

File: vars/rollingUpgradePipeline.groovy [hunk 5]
```diff
@@ -15,6 +15,12 @@ def call(Map pipelineParams) {
             SCT_GCE_PROJECT = "${params.gce_project}"
         }
         parameters {
+            string(defaultValue: '',
```
Comment (vponomaryov) [inline vars/rollingUpgradePipeline.groovy:18]: What happens when the value is empty string?

File: jenkins-pipelines/rolling-upgrade-azure-image.jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#!groovy
+
+// trick from https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/43
+def lib = library identifier: 'sct@snapshot', retriever: legacySCM(scm)
+
+rollingUpgradePipeline(
+    backend: 'azure',
+    azure_region_name: 'eastus',
+    azure_instance_type_db: 'Standard_L8s_v2',
```
Comment (fgelcer) [inline jenkins-pipelines/rolling-upgrade-azure-image.jenkinsfile:9]: shouldn't these parameters go into a configuration file, other than hardcoded in the `jenkinsfile`?

File: vars/createSctRunner.groovy (hunk 1 above)
Comment (fgelcer) [reply vars/createSctRunner.groovy:30]: i don't see it moved... or i don't really understand where it was meant to...

Comment (KnifeyMoloko) [reply vars/createSctRunner.groovy:29]: It was meant to kept the blocks related to `availability_zone` together and apart from the `region_zone` stuff

File: vars/rollingUpgradePipeline.groovy (hunk 5 above)
Comment (KnifeyMoloko) [reply vars/rollingUpgradePipeline.groovy:18]: Not much if the user provides a base version param. If neither base version or image params are given, the job will fail (without a verbose and clear error message). We probably should some validation logic more generally.

File: jenkins-pipelines/rolling-upgrade-azure-image.jenkinsfile (hunk 6 above)
Comment (KnifeyMoloko) [reply jenkins-pipelines/rolling-upgrade-azure-image.jenkinsfile]: Fixed (moved most of the config items to a new config yaml file)

File: vars/rollingUpgradePipeline.groovy (hunk 5 above)
Comment (fgelcer) [reply vars/rollingUpgradePipeline.groovy:18]: it happens anyway with all rolling upgrade, regardless the new one being added by this PR...

**Comment by fgelcer** [comment]
ping @vponomaryov @fruch @soyacz

File: vars/rollingUpgradePipeline.groovy (hunk 5 above)
Comment (vponomaryov) [reply vars/rollingUpgradePipeline.groovy:18]: thnx

**Review by vponomaryov — APPROVED** [review]

**Comment by fgelcer** [comment]
@KnifeyMoloko , can't cleanly backport, so please send a PR to `branch-5.1` (and perhaps `2022.2`)

**Pull request merged**